server-adapters/resteasy-reactor-netty/src/main/java/org/jboss/resteasy/plugins/server/reactor/netty/ChunkOutputStream.java

```diff
@@ -0,0 +1,186 @@
+package org.jboss.resteasy.plugins.server.reactor.netty;
+
+import org.jboss.resteasy.spi.AsyncOutputStream;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+import reactor.core.publisher.MonoProcessor;
+import reactor.netty.http.server.HttpServerResponse;
+import reactor.util.function.Tuple2;
+import reactor.util.function.Tuples;
+
+import java.io.IOException;
+import java.time.Duration;
+import java.util.Arrays;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.atomic.AtomicReference;
+
+/**
+ * This is the output stream leveraged by {@link
+ * ReactorNettyHttpResponse#getOutputStream}.  It provides the heavy lifting
+ * for actually transfering the bytes written by RestEasy to a {@link
+ * Flux<byte[]>}, which is what reactor-netty works with.  Most of the heavy
+ * lifting occurs in {@link #asyncWrite(byte[], int, int)}.
+ */
+public class ChunkOutputStream extends AsyncOutputStream {
+
+   private static final Logger log = LoggerFactory.getLogger(ChunkOutputStream.class);
+
+   private static final boolean COMPLETED_SIGNAL = true;
+
+   private long totalBytesSent;
+
+   /**
+    * Allows the sending of bytes to the client.
+    */
+   private final HttpServerResponse response;
+
+   /**
+    * A signal back to the 'main' Flux that writing bytes back to the client has finished.
+    */
+   private final MonoProcessor<Void> completionMono;
+
+   /**
+    * A sink that will be used to bridge bytes from RestEasy (i.e. this class)
+    * to the `Flux<byte[]>` used in sendByteArray on {@link #response}.
+    */
+   private final AtomicReference<EventListener> listener = new AtomicReference<>();
+
+   /**
+    * This is used to stop the sending of bytes until the sink and {@link #listener} used
+    * with {@link #out} has been fully established.
+    */
+   private final CompletableFuture<Boolean> sinkCreated = new CompletableFuture<>();
+
+   /**
+    * Indicates we've started sending bytes.
+    */
+   private boolean started; // want to eliminate this
+
+   // TODO The use of this is very questionable!  I'm just throwing
+   // this in at the very end before I try something radically different
+   // for the entire adapter.
+   private Duration timeout;
+
+   /**
+    * The {@link Flux<byte[]>} that is fed into {@link HttpServerResponse#sendByteArray}.
+    */
+   private Flux<Tuple2<byte[], CompletableFuture<Boolean>>> out = Flux.create(sink -> {
```

Comment:
Looks like this may lead to a OOM if there is one response that's really huge.  I am not sure if other `OverflowStrategy` options would work though..  except may be erroring..  Maybe we can keep `BUFFER`; however, then do a check somewhere on the size and give an error if it is beyond buffer size..

Reply:
You are worried about the case where this no downstream demand, right?  Anyhow, there is concern about OOM.  I believe the netty adapter restricts payload size.